crates/modelardb_server/src/storage/uncompressed_data_manager.rs

```diff
@@ -336,182 +373,232 @@ impl UncompressedDataManager {
                 univariate_id
             );
 
-            buffer.insert_data(self.current_batch_index, timestamp, value);
+            buffer.insert_data(current_batch_index, timestamp, value);
             self.active_uncompressed_data_buffers
                 .insert(univariate_id, buffer);
         }
 
-        Ok(None)
-    }
+        // Transfer the full buffer to the compressor.
+        if buffer_is_full {
+            debug!(
+                "Buffer for {} is full, moving it to the channel of finished buffers.",
+                univariate_id
+            );
 
-    /// Remove the oldest finished [`UncompressedDataBuffer`] from the queue and return it. Returns
-    /// [`None`] if there are no finished [`UncompressedDataBuffers`](UncompressedDataBuffer) in the
-    /// queue.
-    pub(super) async fn finished_data_buffer(&mut self) -> Option<Box<dyn UncompressedDataBuffer>> {
-        if let Some(uncompressed_data_buffer) = self.finished_uncompressed_data_buffers.pop_front()
-        {
-            // Add the memory size of the removed buffer back to the remaining bytes and record the change.
-            self.memory_pool
-                .free_uncompressed_memory(uncompressed_data_buffer.memory_size());
-            self.used_uncompressed_memory_metric
-                .append(-(uncompressed_data_buffer.memory_size() as isize), true);
+            // unwrap() is safe as this is only reachable if the buffer exists in the HashMap.
+            let full_uncompressed_in_memory_data_buffer = self
+                .active_uncompressed_data_buffers
+                .remove(&univariate_id)
+                .unwrap()
+                .1;
 
-            // Since the buffer was potentially on disk, record the potential change to the used disk space.
-            self.used_disk_space_metric
-                .write()
-                .await
-                .append(-(uncompressed_data_buffer.disk_size() as isize), true);
+            let finished_uncompressed_data_buffer: Box<dyn UncompressedDataBuffer> =
+                if self.memory_pool.remaining_uncompressed_memory_in_bytes() <= 0 {
+                    debug!(
+                        "Over allocated memory for uncompressed data, spilling finished buffer."
+                    );
 
-            Some(uncompressed_data_buffer)
-        } else {
-            None
+                    let uncompressed_on_disk_data_buffer = self
+                        .spill_finished_buffer(full_uncompressed_in_memory_data_buffer)
```

Comment:
Unfortunately I was unable to find a way to spill old buffers while preserving their order as they are transferred between threads using channels to block threads until a message arrives. I thought about spilling buffers that are not full, but the use case where many univariate time series are ingested at the same time would just immediately require them to be read from disk or new buffers to be allocated when the next batch of data points arrive. Many univariate time series is unfortunately a well known problem for TSMSs that manage time series as univariate time series  to handle, e.g., one of the core problems of InfluxDB 1/2 that InfluxDB 3 was designed to solve was to [support infinite cardinality](https://www.influxdata.com/blog/the-plan-for-influxdb-3-0-open-source/). So the current implementation of `UncompressedDataManager` tries to optimize for the RES use case (a reasonable number of very long time series) and  balance not over allocating way too much while still having buffers available for ingesting. However, I am sure it can be improved, e.g., maybe it is worth also spilling unfinished buffers that are no longer used despite them not containing a full set of data points, so I created a task to think about how the memory management of `UncompressedDataManager` can be improved with a link to this disscussion.